--- Day 11: Cosmic Expansion ---
You continue following signs for "Hot Springs" and eventually come across an observatory. The Elf within turns out to be a researcher studying cosmic expansion using the giant telescope here.

He doesn't know anything about the missing machine parts; he's only visiting for this research project. However, he confirms that the hot springs are the next-closest area likely to have people; he'll even take you straight there once he's done with today's observation analysis.

Maybe you can help him with the analysis to speed things up?

The researcher has collected a bunch of data and compiled the data into a single giant image (your puzzle input). The image includes empty space (.) and galaxies (#). For example:

...#......
.......#..
#.........
..........
......#...
.#........
.........#
..........
.......#..
#...#.....
The researcher is trying to figure out the sum of the lengths of the shortest path between every pair of galaxies. However, there's a catch: the universe expanded in the time it took the light from those galaxies to reach the observatory.

Due to something involving gravitational effects, only some space expands. In fact, the result is that any rows or columns that contain no galaxies should all actually be twice as big.

In the above example, three columns and two rows contain no galaxies:

   v  v  v
 ...#......
 .......#..
 #.........
>..........<
 ......#...
 .#........
 .........#
>..........<
 .......#..
 #...#.....
   ^  ^  ^
These rows and columns need to be twice as big; the result of cosmic expansion therefore looks like this:

....#........
.........#...
#............
.............
.............
........#....
.#...........
............#
.............
.............
.........#...
#....#.......
Equipped with this expanded universe, the shortest path between every pair of galaxies can be found. It can help to assign every galaxy a unique number:

....1........
.........2...
3............
.............
.............
........4....
.5...........
............6
.............
.............
.........7...
8....9.......
In these 9 galaxies, there are 36 pairs. Only count each pair once; order within the pair doesn't matter. For each pair, find any shortest path between the two galaxies using only steps that move up, down, left, or right exactly one . or # at a time. (The shortest path between two galaxies is allowed to pass through another galaxy.)

For example, here is one of the shortest paths between galaxies 5 and 9:

....1........
.........2...
3............
.............
.............
........4....
.5...........
.##.........6
..##.........
...##........
....##...7...
8....9.......
This path has length 9 because it takes a minimum of nine steps to get from galaxy 5 to galaxy 9 (the eight locations marked # plus the step onto galaxy 9 itself). Here are some other example shortest path lengths:

Between galaxy 1 and galaxy 7: 15
Between galaxy 3 and galaxy 6: 17
Between galaxy 8 and galaxy 9: 5
In this example, after expanding the universe, the sum of the shortest path between all 36 pairs of galaxies is 374.

Expand the universe, then find the length of the shortest path between every pair of galaxies. What is the sum of these lengths?

In [2]:
with open('input11.txt') as file:
    data = [l.rstrip() for l in file]

def flip(table):
    flipped = [[] for _ in table[0]]
    for line in table:
        for ix, c in enumerate(line):
            flipped[ix].append(c)
    for ix, row in enumerate(flipped):
        flipped[ix] = ''.join(row)
    return flipped

def expand_rows(table):
    exp = []
    for line in table:
        exp.append(line)
        if '#' not in line:
            exp.append(line)
    return exp

def expand_all(table):
    return flip(expand_rows(flip(expand_rows(table))))

In [3]:
sample_str = '''...#......
.......#..
#.........
..........
......#...
.#........
.........#
..........
.......#..
#...#.....'''
sample = [l for l in sample_str.splitlines()]
exp_sample = expand_all(sample)

In [ ]:
exp = expand_all(data)
galaxies = []
for i, line in enumerate(exp):
    for j, pos in enumerate(line):
        if pos == '#':
            galaxies.append((i,j))

sum_short = 0
for ix, gal1 in enumerate(galaxies):
    for gal2 in galaxies[ix+1:]:
        dist = abs(gal1[0] - gal2[0]) + abs(gal1[1] - gal2[1])
        sum_short += dist
sum_short

9563821

Your puzzle answer was 9563821.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---
The galaxies are much older (and thus much farther apart) than the researcher initially estimated.

Now, instead of the expansion you did before, make each empty row or column one million times larger. That is, each empty row should be replaced with 1000000 empty rows, and each empty column should be replaced with 1000000 empty columns.

(In the example above, if each empty row or column were merely 10 times larger, the sum of the shortest paths between every pair of galaxies would be 1030. If each empty row or column were merely 100 times larger, the sum of the shortest paths between every pair of galaxies would be 8410. However, your universe will need to expand far beyond these values.)

Starting with the same initial image, expand the universe according to these new rules, then find the length of the shortest path between every pair of galaxies. What is the sum of these lengths?

In [5]:
exp_rows = [ix for ix, row in enumerate(data) if '#' not in row]
exp_cols = [ix for ix, col in enumerate(zip(*data)) if '#' not in col]
exp_rows, exp_cols

([50, 51, 55, 57, 64, 71, 73, 77, 80],
 [59, 60, 65, 68, 81, 83, 89, 91, 104, 115])

In [7]:
galaxies = []
for i, line in enumerate(data):
    for j, pos in enumerate(line):
        if pos == '#':
            galaxies.append((i,j))

In [8]:
sum_short = 0
for ix, gal1 in enumerate(galaxies):
    for gal2 in galaxies[ix+1:]:
        i1, j1 = gal1
        i2, j2 = gal2
        empties = sum([1 for r in exp_rows if min(i1, i2) < r < max(i1,i2)]) + sum([1 for c in exp_cols if min(j1, j2) < c < max(j1,j2)])
        dist = abs(gal1[0] - gal2[0]) + abs(gal1[1] - gal2[1]) + empties * (1000 ** 2 - 1)
        sum_short += dist
sum_short

827009909817

That's the right answer! You are one gold star closer to restoring snow operations.

You have completed Day 11! You can [Share] this victory or [Return to Your Advent Calendar].